Improved handling of win32 abs paths when executing a generator by path.

mykolav opened the pull request:
Hi!

This pull request makes it possible to say:
```
yo D:\src\projects\vscode-generator-code\generators\app
```

Useful for debugging of generators.

Thank you,
Mykola.

SBoudrias commented:
I'm not sure I see how this patch is fixing the issue in all cases - and across OS.

Could you explain why you think your code fixes the issue?

mykolav commented:
Hi SBoudrias,
Sure.

**1)** Please notice I changed the pull request's description.
It was `yo code D:\src\projects\vscode-generator-code\generators\app` but it should have been `yo D:\src\projects\vscode-generator-code\generators\app`
`code` - is indeed the name of the generator I was trying to debug.

**2)** Let's take a look at a portion of code from Environment.prototype.get = function get(namespace) with the pull request merged:
```
var parts = namespace.split(':');
  var maybePath = _.last(parts);
  if (parts.length > 1 && /[\/\\]/.test(maybePath)) {
    parts.pop();

    // We also want to remove the drive letter on windows
    if (maybePath.indexOf('\\') >= 0 && _.last(parts).length === 1) {
      parts.pop();
    }
```
As I understand, the code above is supposed to remove the `path` portion from the `namespace` parameter's value that potentially contains `namespace:path`

Let's consider a string of the form `a:\path` (where `a` is a drive letter). It is a valid absolute path in Win32 family of OS-es.

But the `parts` collection split from `a:\path` is empty at a point where the `parts` are joined.
How come? First, from `\path` is removed from `a:\path` on assumption that `a` is a namespace name.
Second, the `We also want to remove the drive letter on windows...` conditional is executed and it removes the drive letter.

```
    if (parts.length > 0) {
      namespace = parts.join(':');
    }
```
Making sure, `parts` is not empty before overwriting the `namespace`'s original value helps handle Win32-type absolute path properly.

**3)** How does this patch work across other OS-es? I believe nothing is broken or fixed for OS-es with POSIX paths. 
Admittedly though:

**3 a)**  Without this patch, `parts` split from a string of the form `n:path/to/something` (where `n` is a single letter namespace) are emptied. And as a result the `namepsace` is overwritten with empty string when joining `parts` back.

**3 b)** With this patch, `parts` split from a string of the form `n:path/to/something` (where `n` is a single letter namespace) are emptied. And as a result the original value of `namepsace` is preserved.

Apparently, both approaches fail if the legacy yeoman-generator `#hookFor()` supplies the code with `namespace` of the form `n:path/to/generator`.
But how likely it is a single letter namespace will ever by encounter?

Thank you,
Mykola.

SBoudrias commented on lib/environment.js:
```diff
@@ -253,7 +253,12 @@ Environment.prototype.get = function get(namespace) {
       parts.pop();
     }
 
-    namespace = parts.join(':');
```
So, I think what we want to do here, is only not overwriting the `namespace` value. - Which is what your code is doing conditionally.

But if we actually only use two different variables names, then it'll just work without the need for a conditional.

How about we rename `namespace` to `cleanedNamespace`?

SBoudrias commented:
Sorry, the review interface is kind of new for me and it looks a little weird. But thanks for taking the time of walking me through the bug. It's now pretty clear in my head what is wrong.

I think the alternative solution presented in the review would simplify the code and reduce chance of causing edge cases.

mykolav commented:
Hi,

Implemented the improvement you suggested in [89e0250](https://github.com/yeoman/environment/pull/69/commits/89e0250218ea7097147995eb5e7e5dd88f8746b9)

Please notice, I named the function's parameter `namespaceOrPath` so instead of `cleanedPath` there is simply the `namespace` variable which is supposed to be `namespaceOrPath` stripped of a potential portion. 
To me it seems the names `namespaceOrPath` and `path` express the purpose of these variables better. But I don't insist on these specific names. If you prefer `cleanedPath` I'll rename.

Adding a test is a great idea. I'll investigate in a day or two.

Thank you,
Mykola.

mykolav commented:
Hello,

Added a test in [88511a0](https://github.com/yeoman/environment/pull/69/commits/88511a0eb2bed1160accb06998fbfeb25ee76719)

Some aspects of the test's code that I would really appreciate a suggesting on:

1. The code creates `sion.stub`-s at the begging of the test-case and `restore`-s the stubs at the end. The stubs are specific to this individual test-case so I did not place them in `beforeEach()` and `afterEach()`. But OTOH, this code seems to distract a reader from the test-case's "core" code.

2. I test all the three stubs in one test-case. Maybe each stub should be tested in its individual test-case?

Thank you,
Mykola.

SBoudrias commented on test/environment.js:
```diff
@@ -431,6 +431,38 @@ describe('Environment', function () {
+      var aliasMethod;
+      var getByPathMethod;
+
+      var absolutePaht = 'C:\\foo\\bar'; 
```
paht -> path

SBoudrias commented on test/environment.js:
```diff
@@ -431,6 +431,38 @@ describe('Environment', function () {
+
+      var absolutePaht = 'C:\\foo\\bar'; 
+
+      try
```
`try {` same with the finally block

SBoudrias commented on test/environment.js:
```diff
@@ -431,6 +431,38 @@ describe('Environment', function () {
+        assert.ok(aliasMethod.calledOnce);
+        assert.ok(aliasMethod.calledWithExactly(""));
+
+        assert.ok(getByPathMethod.calledOnce);
```
In this test, we only care if `getByPathMethod` is called. The previous assertions are not relevant to the feature we're testing.

Also, here you can write `sinon.assert.calledWith(getByPathMethod, absolutePath)` - the assertion error will be more useful if the test ever come to fail.

mykolav commented:
Thank you for reviewing the test-case and the useful tips.

A cleaned-up version of it is in [1ab1474](https://github.com/yeoman/environment/pull/69/commits/1ab1474701b455e22ed64184500646f1fd143e4a)

Please take a look, when you have a chance.

Thank you,
Mykola.

SBoudrias commented:
Thanks a ton! Merging :)
